pkg/vsphere/vm/vm.go

```diff
@@ -272,17 +272,27 @@ func (vm *VirtualMachine) DeleteExceptDisks(ctx context.Context) (*types.TaskInf
 
 	op := trace.FromContext(ctx, "DeleteExceptDisks")
 
-	op.Debugf("Getting list of the devices for VM %q", vm)
-	devices, err := vm.Device(op)
-	if err != nil {
-		return nil, err
-	}
+	firstTime := true
+	err := retry.Do(func() error {
+		op.Debugf("Getting list of the devices for VM %q", vm)
+		devices, err := vm.Device(op)
+		if err != nil {
+			return err
+		}
 
-	disks := devices.SelectByType(&types.VirtualDisk{})
-	op.Debug("Removing devices for VM")
+		disks := devices.SelectByType(&types.VirtualDisk{})
+		if len(disks) > 0 {
+			op.Debugf("Removing devices for VM %q", vm)
+			firstTime = false
+			return vm.RemoveDevice(op, true, disks...)
+		} else if firstTime {
```

Comment:
The differentiation in logging here can easily be inferred based on whether or not a retry has been logged and or the presence of "Removing devices for..." message.

If you keep the differentiated logging, please simplify the code structure to improve readability by splitting the `else if` portion of the clause into a separate `if firsttime { ... } else { ...}`.